browser/net/brave_ad_block_tp_network_delegate_helper.cc

```diff
@@ -56,16 +84,89 @@ void OnShouldBlockAdResult(const ResponseCallback& next_callback,
   DCHECK_CURRENTLY_ON(content::BrowserThread::UI);
   if (ctx->blocked_by == kAdBlocked) {
     brave_shields::DispatchBlockedEvent(
-        ctx->request_url,
-        ctx->render_frame_id, ctx->render_process_id, ctx->frame_tree_node_id,
-        brave_shields::kAds);
+        ctx->request_url, ctx->render_frame_id, ctx->render_process_id,
+        ctx->frame_tree_node_id, brave_shields::kAds);
   }
   next_callback.Run();
 }
 
-void OnBeforeURLRequestAdBlockTP(
+void ShouldBlockAdWithOptionalCname(
+    scoped_refptr<base::SequencedTaskRunner> task_runner,
     const ResponseCallback& next_callback,
-    std::shared_ptr<BraveRequestInfo> ctx) {
+    std::shared_ptr<BraveRequestInfo> ctx,
+    const base::Optional<std::string> cname) {
+  DCHECK_CURRENTLY_ON(content::BrowserThread::UI);
+  task_runner->PostTaskAndReply(
+      FROM_HERE, base::BindOnce(&ShouldBlockAdOnTaskRunner, ctx, cname),
+      base::BindOnce(&OnShouldBlockAdResult, next_callback, ctx));
+}
+
+class AdblockCnameResolveHostClient : public network::mojom::ResolveHostClient {
+ private:
+  mojo::Receiver<network::mojom::ResolveHostClient> receiver_{this};
+  base::OnceCallback<void(base::Optional<std::string>)> cb_;
+
+ public:
+  AdblockCnameResolveHostClient(
+      base::OnceCallback<void(base::Optional<std::string>)> next_callback,
+      std::shared_ptr<BraveRequestInfo> ctx) {
+    cb_ = std::move(next_callback);
+
+    auto* web_contents = GetWebContents(ctx->render_process_id,
+                                        ctx->render_frame_id,
+                                        ctx->frame_tree_node_id);
+    DCHECK(web_contents);
```

Comment:
not sure this will work for service workers

Reply:
@iefremov how might I fix that?

Reply:
just replace with `if`

Reply:
you're correct, I noticed that Twitch has this case occasionally. Should be fixed now.